Support for Hachirumi

Curstantine opened the pull request:
Added basic methods:

- `getMangaDetails`
- `getChapters`
- `getChapterDetails`
- `searchRequest`

ConradWeiser commented on src/Hachirumi/Hachirumi.ts:
```diff
@@ -0,0 +1,187 @@
+        ? JSON.parse(response.data)
+        : response.data;
+
+    let chapterObject = result["chapters"];
```
While in normal circumstance this probably works most of the time, it's (unlikely, but) possible for this to be `undefined` or malformed. If this is the case, line 90 will not be iterable and the error it spits out won't make sense to most normal users.

I suggest that you check that `chapterObject` and `groupObject` are defined before continuing with your loop on line 90.

If they are not defined, throw an error with some nice human readable message stating why.

ConradWeiser commented on src/Hachirumi/Hachirumi.ts:
```diff
@@ -0,0 +1,187 @@
+
+    for (let key in chapterObject) {
+      let metadata = chapterObject[key];
+      for (let groupKey in metadata["groups"]) {
```
Ensure that `metadata["groups"]` is defined before attempting to loop it, throw otherwise!

ConradWeiser commented on src/Hachirumi/Hachirumi.ts:
```diff
@@ -0,0 +1,187 @@
+          createChapter({
+            id: `${key}|${groupKey}|${metadata["folder"]}`, // Moved to this format as it is easier to find the `key, groupkey, folder`.
+            mangaId: result.slug,
+            chapNum: parseInt(key),
```
It is possible for this to be `NaN` - I'd recommend something more fancy to ensure this doesn't screw things up.
`isNaN(parseInt(key)) ? parseInt(key) : -1` <-- At a minimum something like this, although ideally you should probably check this out before you get here

ConradWeiser commented on src/Hachirumi/Hachirumi.ts:
```diff
@@ -0,0 +1,187 @@
+            name: metadata["title"],
+            volume: metadata["volume"],
+            group: groupObject[groupKey],
+            time: new Date(metadata["release_date"][groupKey] * 1000),
```
It's fine on line 99 or 100 that it might be undefined, but if you're attempting to multiply something by 1000, you should ensure that `metadata["release_date"][groupKey]` is defined before attempting to do this.

ConradWeiser commented on src/Hachirumi/Hachirumi.ts:
```diff
@@ -0,0 +1,187 @@
+        ? JSON.parse(response.data)
+        : response.data;
+
+    let chapterObject = result["chapters"];
```
Make sure this is defined before attempting to use it. This appears to be a value you can't afford to be undefined.

ConradWeiser commented on src/Hachirumi/Hachirumi.ts:
```diff
@@ -0,0 +1,187 @@
+      id: chapterId,
+      longStrip: false, // Not implemented.
+      mangaId: mangaId,
+      pages: chapterObject[chapterKey]["groups"][groupKey].map((ext: string) =>
```
Ensure that `chapterObject[chapterKey]["groups"][groupKey]` is defined before attempting to use it. 
Sidenote: I like your usage of map quite a bit!

ConradWeiser reviewed (commented):
All in all, you have quite the nice coding convention! 
Because of the nature of sources being used, if Hachimuri is to ever change how their site is rendered, it's usually a very good idea to ensure that the source is bulletproof. 

If there is a case where it CAN fail, it should log something which makes sense to help identify exactly where it has failed. This is also useful for users who may have internet problems, trying to debug what's going on. The attached suggestions mostly target this 'always have an error case' ideology.

When you `throw` an error, users see what that error was on a snackbar popup. It should be some kind of text where if it was reported to you, or anyone else attempting to maintain the source, that we'd be able to identify what was happening immediately. Fortunately, this source seems to be a bit on the simple side, so this shouldn't be TOO much of an issue, but it's good form.

I'll make a follow up review later once I manage to get this tested on a device! (Though I suspect it works great on a device, provided that you've PRed this) - Thanks!

Curstantine commented:
Thanks! I'll look into it tomorrow.

Curstantine commented:
Added a new file called `functions.ts` it holds type checking and error handling for response fields.

ConradWeiser reviewed (dismissed):
There is not any standard available right now, the method you've used works quite nicely! 
I'm not catching any other issues.

I have not tested it on an app with the error handling, but I've checked it on your previous commit finding no glaring problems!

We'll need one more reviewer to take a look. Thanks for your work on this!

Curstantine commented:
1. [Added a prettier file to remove semi-colons.](https://discord.com/channels/673606787290759230/705193595715452958/859373321610592306)

2. Latest Updates section with 3 slots. (I'm not sure about the IDs and such).

3. Moved from `let => const` for unused mutables.

4. Added status code based error handling for fetching.

Curstantine commented:
Made a big mess.
